### PR TITLE
[Draft] TopK Fusion to JAX

### PR DESCRIPTION
**Pull request by mingxu1067**
# Description

- Adding TopK fusion to JAX for both forward and backward.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Adding TopK fusion to JAX for both forward and backward.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/jax/router.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
+# Copyright (c) 2022-2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (phu0ngng) [inline transformer_engine/jax/router.py:1]: Hi @mingxu1067,

Could you rename this file to moe.py, as we will host the permutation ops here as well?

File: transformer_engine/jax/cpp_extensions/router.py [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+    "map_score_function",
+]
+
+
+def map_score_function(score_function: str) -> int:
+    score_function_map = {"sigmoid": 0, "softmax": 1}
+    assert (
+        score_function in score_function_map
+    ), f"score_function must be 'sigmoid' or 'softmax', got {score_function}"
+    return score_function_map[score_function]
```
Comment (phu0ngng) [inline transformer_engine/jax/cpp_extensions/router.py:28]: Hi, should we have an enum instead?

File: transformer_engine/jax/cpp_extensions/router.py [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+        te_fused_topk_with_score_function_forward abstract
+        """
+        dtype = dtypes.canonicalize_dtype(logits_aval.dtype)
+        assert len(logits_aval.shape) == 3  # (batch, seqlen, num_experts)
+
+        probs_aval = logits_aval.update(shape=logits_aval.shape, dtype=dtype)
+        routing_map_aval = logits_aval.update(shape=logits_aval.shape, dtype=jnp.bool_)
+        intermediate_output_aval = logits_aval.update(shape=logits_aval.shape, dtype=dtype)
+
+        return probs_aval, routing_map_aval, intermediate_output_aval
```
Comment (phu0ngng) [inline transformer_engine/jax/cpp_extensions/router.py:71]: Hi, may I learn what the intermediate_output is for?

File: transformer_engine/jax/cpp_extensions/router.py [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+            score_function,
+            mesh,
+            result_types,
+        )
+
+        prefix = "Router_"
+        logits_spec = (prefix + "batch", prefix + "seqlen", prefix + "experts")
+        expert_bias_spec = (prefix + "experts",)
+
+        output_spec = (prefix + "batch", prefix + "seqlen", prefix + "experts")
```
Comment (phu0ngng) [inline transformer_engine/jax/cpp_extensions/router.py:225]:
```suggestion
        output_spec = logits_spec
```

File: transformer_engine/jax/csrc/extensions/router.cpp [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+#include <vector>
+
+#include "../extensions.h"
+
+namespace transformer_engine {
+namespace jax {
+
+// Score function mapping: sigmoid=0, softmax=1
+constexpr int kScoreFunctionSigmoid = 0;
+constexpr int kScoreFunctionSoftmax = 1;
```
Comment (phu0ngng) [inline transformer_engine/jax/csrc/extensions/router.cpp:19]: CPP Enum

**Review by phu0ngng — COMMENTED** [review]
Hi Ming, could you also add a unit test for this topK? Thanks!

File: transformer_engine/jax/cpp_extensions/router.py (hunk 3 above)
Comment (tdophung) [reply transformer_engine/jax/cpp_extensions/router.py:71]: it is to be saved for backward pass. intermediate_outputs is just the softmax(logits) or sigmoid(logits) pre/post topk depending on whether use_pre-softmax = True/False

**Comment by phu0ngng** [comment]
Close as the work has been done via #2711.
